// ngx-scanner-qrcode/esm2022/lib/ngx-scanner-qrcode.service.mjs
import { Injectable } from '@angular/core';
import { AsyncSubject } from 'rxjs';
import { AS_COMPLETE, COMPRESS_IMAGE_FILE, FILES_TO_SCAN } from './ngx-scanner-qrcode.helper';
import * as i0 from "@angular/core";
class NgxScannerQrcodeService {
    /**
     * loadFiles
     * @param files
     * @param percentage
     * @param quality
     * @returns
     */
    loadFiles(files = [], percentage, quality) {
        const as = new AsyncSubject();
        COMPRESS_IMAGE_FILE(files, percentage, quality).then((_files) => {
            Promise.all(Object.assign([], _files).map((m) => this.readAsDataURL(m))).then((img) => AS_COMPLETE(as, img)).catch((error) => AS_COMPLETE(as, null, error));
        });
        return as;
    }
    /**
     * loadFilesToScan
     * @param files
     * @param config
     * @param percentage
     * @param quality
     * @returns
     */
    loadFilesToScan(files = [], config, percentage, quality) {
        return FILES_TO_SCAN(files, config, percentage, quality);
    }
    /**
     * readAsDataURL
     * @param file
     * @returns
     */
    readAsDataURL(file) {
        /** drawImage **/
        return new Promise((resolve, reject) => {
            const fileReader = new FileReader();
            fileReader.onload = () => {
                const objectFile = {
                    name: file.name,
                    file: file,
                    url: URL.createObjectURL(file)
                };
                resolve(objectFile);
            };
            fileReader.onerror = (error) => reject(error);
            fileReader.readAsDataURL(file);
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.3", ngImport: i0, type: NgxScannerQrcodeService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.0.3", ngImport: i0, type: NgxScannerQrcodeService, providedIn: 'root' }); }
}
export { NgxScannerQrcodeService };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.3", ngImport: i0, type: NgxScannerQrcodeService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }] });
