// ngx-scanner-qrcode/esm2022/lib/ngx-scanner-qrcode.module.mjs
import { NgModule } from '@angular/core';
import { NgxScannerQrcodeService } from './ngx-scanner-qrcode.service';
import { NgxScannerQrcodeComponent } from './ngx-scanner-qrcode.component';
import * as i0 from "@angular/core";
class NgxScannerQrcodeModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.3", ngImport: i0, type: NgxScannerQrcodeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.0.3", ngImport: i0, type: NgxScannerQrcodeModule, declarations: [NgxScannerQrcodeComponent], exports: [NgxScannerQrcodeComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.0.3", ngImport: i0, type: NgxScannerQrcodeModule, providers: [NgxScannerQrcodeService] }); }
}
export { NgxScannerQrcodeModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.3", ngImport: i0, type: NgxScannerQrcodeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxScannerQrcodeComponent],
                    exports: [NgxScannerQrcodeComponent],
                    providers: [NgxScannerQrcodeService],
                }]
        }] });
